### ISSUE TITLE
Replace`asserts` with `require`

### ISSUE THREAD

**Issue by emizzle** [codex-storage/codex-contracts-eth#149]
See comment https://github.com/codex-storage/codex-contracts-eth/pull/144#discussion_r1710864678.

**Comment by markspanbroek** [comment]
There is a semantic difference between `require()` and `assert()`. Preconditions such as requirements on a function's input are handled by `require()`, Things that a programmer thinks should never happen are handled by `assert()`.
This semantic difference is for example [used by echidna](https://github.com/crytic/building-secure-contracts/blob/master/program-analysis/echidna/basic/assertion-checking.md#when-and-how-to-use-assertions) for fuzzing. They use `require()` to ensure that fuzzed input values match expected ranges, and then check that assertions don't happen.

Also I expect there to be a good reason why asserts eat up all the gas; maybe to ensure that exceptions are not caught and then handled elsewhere. It could be a security property that ensures that execution really stops.

**Comment by 0x-r4bbit** [comment]
@markspanbroek you're not wrong, and we're probably entering the space of bike shedding here.
I'd extend the definition of where/when to use `require` to "use for conditions that depend on **external factors** the code doesn't control", which is user inputs as you've described, but also external calls like transferring tokens.
This also allows for providing proper custom errors (which asserts don't allow for). 

Assertions are more appropriate for internal checks and typically used for cases when there's a serious internal error to your code.

Ultimately it's your decision, but I at least wanted to bring it up.

**Comment by markspanbroek** [comment]
Ah, good point @0x-r4bbit, the call to the ERC20 contract can indeed be seen as an external call which the code doesn't control. Using `require()` does make sense there. I read the description of this issue as "we're going to replace all asserts with requires".
